library_management_system_downloader/copy_and_compress_database_and_extract_tsv.py

```diff
@@ -0,0 +1,119 @@
+
+# =============================================================================
+# Settings
+# =============================================================================
+
+sqlite_database_location = 'library_coverage_xml_and_fulltext_indicators.db'
+
+data_directory = 'data'
+
+compressed_database_copy_name = (
+        'library_coverage_xml_and_fulltext_indicators'
+        '.db.xz')
+
+tsv_output_file_name = 'library_coverage_xml_and_fulltext_indicators.tsv.xz'
+
+# =============================================================================
+# Create additional settings
+# =============================================================================
+
+logging.basicConfig(level=logging.INFO)
+
+# =============================================================================
+# Set up a connection to (and tables within) the SQLite database:
+# =============================================================================
+
+# Initialize the database:
+Base = automap_base()
+sql_engine = create_engine(f'sqlite:///{sqlite_database_location}')
+Base.prepare(sql_engine, reflect=True)
+
+dois_table = Base.classes.dois_table
+library_holdings_data = Base.classes.library_holdings_data
+
+# From the sqlalchemy documentation
+# (http://docs.sqlalchemy.org/en/latest/orm/session_basics.html)
+# Create a configured "session":
+Session = sessionmaker(bind=sql_engine)
+sql_session = Session()
+
+# =============================================================================
+# Create the data directory if it doesn't already exist
+# =============================================================================
+
+Path(data_directory).mkdir(parents=True, exist_ok=True)
+
+# =============================================================================
+# Extract data from the databse, and save that as a TSV
+# =============================================================================
+
+# We'll first define a subquery to get just the latest full text information
+# for each DOI:
+dataset_max_timestamp_for_each_doi_query = sql_session.query(
+        func.max(library_holdings_data.primary_key)).group_by(
+                library_holdings_data.doi_foreign_key)
+
+# str(dataset_max_timestamp_for_each_doi_query)  # See the SQL being used.
+# list(dataset_max_timestamp_for_each_doi_query)  # See the list of max
+# primary keys.
+
+# We'll now get the DOI and full-text indicator based on the subquery above:
+dataset_join_query = sql_session.query(
+        dois_table.doi,
+        # library_holdings_data.timestamp,  # Uncomment to also get the
+        # timestamp of the DOI fulltext query
+        # library_holdings_data.xml_response,  # Uncomment to also get the
+        # full XML response.
+        library_holdings_data.full_text_indicator) \
+        .join(library_holdings_data) \
+        .group_by(dois_table.doi) \
+        .filter(library_holdings_data.primary_key.in_(
+                dataset_max_timestamp_for_each_doi_query))
```

Comment:
I expect that there is a way to sort by date within a group and just take the first row. Not a huge concern for me but this method seems a bit overkill?

Reply:
I need to sit down tomorrow and experiment with this -- I didn't know that SQL will allow you to sort first and then do a `GROUP BY` on those sorted rows in one query (am I reading your suggestion correctly, that that's what you're suggesting?). This approach translates out of `sqlalchemy` into a standard approach I've seen, to use a subquery to get the IDs one wants first, and then to return just those. Having said that, line 81 is possibly superfluous(? I'm not sure yet).

I think that if we loaded the whole database into a `pandas` dataframe, it would be more straightforward to do what you suggested. But I remember you mentioning earlier that you'd like the code to work at larger scales, too. So keeping things within the SQLite database and then returning just the already-filtered rows seems intuitively to me to be more in line with that goal. Does that all jibe with your understanding?

Reply:
My suggestion (which I think you understand) is:

1. sort by date
2. groupby
3. take first row within group

Of course, 1 and 2 can likely be swapped. My intuition is that this method is possible in SQL / sqlalchemy and more straightforward and efficient than the current approach. However, if my suggestion seems difficult and the current approach is working, then we can continue with it.

Reply:
Thanks for clarifying, so that I'm sure we're on the same page! : )

I remember more clearly now why I used the approach I did. I'm writing all this out because I initially agreed with your intuition that this can be simpler, but then started writing out the steps I'd need to take, and ended up thinking that the current approach is actually *necessary* (though I'd be interested to be shown to be wrong!). I want to make sure my thinking is clear about it (so that we can both review it, and so that it's easier to remember if needed in the future):

As I understand, SQL requires that `GROUP BY` statements come before `ORDER BY` statements. Since we want to be sure of the order of the rows before grouping them, we'd need to use a subquery to order them first, like this:
```sql
SELECT ...
FROM (
    SELECT ...
    FROM ...
    ORDER BY ...
)
GROUP BY ...
```
which is basically what the `sqlalchemy` statement does: running `print(dataset_join_query.statement)`, the `sqlalchemy` code "compiles" into the following:
```sql
SELECT dois_table.doi, library_holdings_data.full_text_indicator 
FROM dois_table
JOIN library_holdings_data 
ON dois_table.database_id = library_holdings_data.doi_foreign_key 
WHERE library_holdings_data.primary_key IN (
    SELECT max(library_holdings_data.primary_key) AS max_1 
    FROM library_holdings_data
    GROUP BY library_holdings_data.doi_foreign_key
)
GROUP BY dois_table.doi
```
But, in this case, the `SORT` is implicit in the fact that the highest `primary_key` for each group will be the newest record for that group. So, we could change this into
```sql
SELECT MAX(dois_table.doi), library_holdings_data.full_text_indicator 
FROM dois_table
JOIN library_holdings_data 
ON dois_table.database_id = library_holdings_data.doi_foreign_key 
GROUP BY dois_table.doi
```
but then there would be no guarantee that the two columns being returned would always be from the same row.

So, we could go back to using a subquery, which I started to write out... and then ended up writing out the same basic approach as in the current `sqlalchemy` code. That seems clearer to me, as I write it out; does it to you, too?

Reply:
Ah looks like you're right that there's no straightforward way to do this. See https://stackoverflow.com/a/7745635/4651668. In fact, it's so common that there's a [`greatest-n-per-group`](https://stackoverflow.com/questions/tagged/greatest-n-per-group) tag!

Anyways, go with whatever you think is best... you know SQL better than I!

Reply:
Ha, cheers! : P I'd prefer sticking with this way, then. This is my first time using `sqlalchemy`, and it reads here a little bit more unwieldy than just the straight SQL, but it's consistent with the rest of the code.